[WGSL] Don't rewrite complex constants

tadeuzagallo opened the pull request:
#### dd455ea08bb0f0467f73e9d8fa271b2bf5c65dba
<pre>
[WGSL] Don&apos;t rewrite complex constants
<a href="https://bugs.webkit.org/show_bug.cgi?id=257486">https://bugs.webkit.org/show_bug.cgi?id=257486</a>
rdar://110005210

Reviewed by Dan Glastonbury.

The original implementation assumed we would completely remove every constant
declaration, but upon further consideration, it doesn&apos;t seem like a great idea
to replace every reference to a constant with its value when the value is a
non-trivial value. i.e. it makes sense to replace literals, but probably not
with vectors, arrays, matrices and structs.

* Source/WebGPU/WGSL/ConstantRewriter.cpp:
(WGSL::ConstantRewriter::materialize):
* Source/WebGPU/WGSL/tests/lit.cfg:
* Source/WebGPU/WGSL/tests/valid/local-constant-vector.wgsl: Added.

Canonical link: <a href="https://commits.webkit.org/264723@main">https://commits.webkit.org/264723@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d5339828a0897da98e037deab2d8546242cde049

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8337 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8630 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8847 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10005 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8402 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10618 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8541 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11268 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8483 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9539 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7559 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10153 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6837 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15192 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7958 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7771 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11119 "7 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8242 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6716 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7531 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2044 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11741 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7985 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg reviewed (approved):
:shipit:

Let's land this to unblock the dependent work and see where the refactoring tell us we need to go.

tadeuzagallo committed:
[WGSL] Don't rewrite complex constants
https://bugs.webkit.org/show_bug.cgi?id=257486
rdar://110005210

Reviewed by Dan Glastonbury.

The original implementation assumed we would completely remove every constant
declaration, but upon further consideration, it doesn't seem like a great idea
to replace every reference to a constant with its value when the value is a
non-trivial value. i.e. it makes sense to replace literals, but probably not
with vectors, arrays, matrices and structs.

* Source/WebGPU/WGSL/ConstantRewriter.cpp:
(WGSL::ConstantRewriter::materialize):
* Source/WebGPU/WGSL/tests/lit.cfg:
* Source/WebGPU/WGSL/tests/valid/local-constant-vector.wgsl: Added.

Canonical link: https://commits.webkit.org/264723@main
